llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp

```diff
@@ -1499,6 +1499,12 @@ std::optional<APInt> Vectorizer::getConstantOffset(Value *PtrA, Value *PtrB,
 
   // Try to compute B - A.
   const SCEV *DistScev = SE.getMinusSCEV(SE.getSCEV(PtrB), SE.getSCEV(PtrA));
+  if (DistScev->isZero()) {
+    // A load in the chain is dependent on another load in the chain, and
+    // attempting to vectorize this chain would create a cycle.
```

Comment:
I'm surprised the result in this situation is a 0. In the test case, how is %gep.x.2 - %gep.x.1 0?

Reply:
The SCEV expression corresponding to `%gep.x.2` is computed as the SCEVUnknown `%x`, which I think is correct. Did you expect an AddExpr?

It took me some time to figure out the fix, although it's quite a simple one in its final form: I think the patch correctly identifies the root cause.

Reply:
So the expression is unknown - unknown = 0? That is surprising behavior. I would expect to filter out either half being unknown

Reply:
that is, I would expect to bail on getSCEV(ptrB) == unknown and if getSCEV(ptrB) == unknown

Reply:
Difference between the _same_ two SCEVUnknowns is zero, and I've seen several instances of this in the tree. Not different SCEVUnknowns. In the case of different SCEVUnknowns, the difference wouldn't have a constant value anyway. No, I don't think we should bail on two unknowns.